### PR TITLE
github oauth and username creation

### PR DESCRIPTION
**Pull request by oliverye7**
PR isnt actually done yet (i think) but i just want someone to look at the code i wrote to roast me for not typing my variables or something silly i forgot :DD

### CODE REVIEW

**Comment by oliverye7** [comment]
stuff that is done:
user can authenticate with github (and stay logged in, via session store)
user can create an in game username
user can log out

**Comment by oliverye7** [comment]
removed 30k lines of code cuz i accidentally included package-lock.json

File: backend/src/api/user.rs [hunk 1]
```diff
@@ -1,15 +1,180 @@
...
-use prost::Message;
+//use log::info;
+use actix_web::{get, post, web, HttpRequest, HttpResponse, Responder};
+use reqwest;
+use serde::{Serialize, Deserialize};
+use sqlx::Pool;
+use std::str::FromStr;
+
+const GITHUB_CLIENT_ID: &str = "d4bd59212a9e47a3ddcd";
+const GITHUB_CLIENT_SECRET: &str = "b5d1af3e2605448d72824627019670c44587be91";
```
Comment (seany4ng) [inline backend/src/api/user.rs:9]: can you give some background/context on how you setup github auth? not familiar with the process but good job

**Review by seany4ng — APPROVED** [review]
too lazy to read through this thoroughly or even pull and test locally but lgtm. phenomenal progress

File: backend/src/api/user.rs (hunk 1 above)
Comment (rahulk29) [inline backend/src/api/user.rs:9]: why did you put secrets in a public repo?

File: backend/src/api/user.rs [hunk 2]
```diff
@@ -1,15 +1,180 @@
...
+    }
+}
+
+#[get("/getUserName")]
+pub async fn get_username(
+    pool: web::Data<Pool<sqlx::Postgres>>,
+    query: web::Query<UserNameQuery>
+) -> impl Responder {
+    let id = &query.id;
+    println!("RAHHHHHHH");
```
Comment (rahulk29) [inline backend/src/api/user.rs:90]: sus

File: backend/src/api/user.rs [hunk 3]
```diff
@@ -1,15 +1,180 @@
...
+    .await
+    {
+        Ok(_record) => {
+            HttpResponse::Ok().json(format!("User {} successfully added", data.username))
+        }
+        Err(e) => HttpResponse::InternalServerError().body(format!("Server Error: {}", e)),
+    }
+}
+
+#[get("/getUserName")]
```
Comment (rahulk29) [inline backend/src/api/user.rs:84]: @rohankk whats the URL naming convention?

File: backend/src/api/user.rs [hunk 4]
```diff
@@ -1,15 +1,180 @@
...
+#[derive(Serialize)]
+pub struct UsernameJsonResponse {
+    username: String,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct UserData {
+    username: String,
+    avatar_url: String,
+    id: i32,
```
Comment (rahulk29) [inline backend/src/api/user.rs:30]: where do you create user IDs? or are these IDs the same as github user IDs?

File: backend/src/api/user.rs (hunk 1 above)
Comment (oliverye7) [reply backend/src/api/user.rs:9]: ohhhhh shucks

**Review by rahulk29 — COMMENTED** [review]
should document the relation between github user data and vimgame user data

File: backend/src/api/user.rs (hunk 4 above)
Comment (oliverye7) [reply backend/src/api/user.rs:30]: id comes from github user IDs, github api call returns a bunch of junk data and i only want to keep username avatar_url and id

File: backend/src/api/challenge.rs [hunk 5]
```diff
@@ -13,6 +12,13 @@ pub struct ChallengeResponse {
     message: String,
 }
 
+#[derive(Deserialize, Serialize)]
+pub struct ChallengeSubmission {
+    challenge_id: i16,
+    user_id: i32,
+    user_keystrokes: Vec<String>,
```
Comment (rahulk29) [inline backend/src/api/challenge.rs:19]: Should this just be a `String`? Or a `Vec<char>`?

File: backend/src/api/user.rs [hunk 6]
```diff
@@ -1,15 +1,196 @@
...
+/*
+ * Updates db with a new username, mapped to a user's id
+*/
+#[post("/setUsername")]
+pub async fn set_username(
+    pool: web::Data<Pool<sqlx::Postgres>>,
+    data: web::Json<UserData>,
+) -> impl Responder {
+    let conn = pool.get_ref();
+    match sqlx::query!("SELECT * FROM users WHERE username = $1;", data.username)
```
Comment (rahulk29) [inline backend/src/api/user.rs:66]: this should be in a transaction, otherwise you are sucsceptible to TOCTTOU issues

**Pull request merged**